src/session/content/message_row/reply.rs

```diff
@@ -0,0 +1,288 @@
+            let obj = self.obj();
+
+            match pspec.name() {
+                "chat" => obj.imp().chat.borrow().clone().to_value(),
+                "reply-id" => obj.imp().reply_id.get().to_value(),
+                "is-loading" => obj.imp().is_loading.get().to_value(),
+                "is-outgoing" => obj.imp().is_outgoing.get().to_value(),
+                "show-thumbnail" => obj.show_thumbnail().to_value(),
+                _ => unimplemented!(),
+            }
+        }
+
+        fn set_property(&self, _id: usize, value: &glib::Value, pspec: &glib::ParamSpec) {
+            let obj = self.obj();
+
+            match pspec.name() {
+                "chat" => obj.set_chat(value.get().unwrap()),
+                "reply-id" => obj.imp().reply_id.set(value.get().unwrap()),
+                "is-outgoing" => obj.imp().is_outgoing.set(value.get().unwrap()),
+                "is-loading" => obj.imp().is_loading.set(value.get().unwrap()),
+                "show-thumbnail" => obj.set_show_thumbnail(value.get().unwrap()),
+                _ => unimplemented!(),
+            }
+        }
+
+        fn constructed(&self) {
+            // do not show thumbnail by default
+            self.show_thumbnail.set(false);
+
+            self.message_thumbnail
+                .connect_notify_local(Some("paintable"), clone!(@weak self as obj => move |thumbnail, _| {
+                    thumbnail.set_visible(thumbnail.paintable().is_some() && obj.show_thumbnail.get())
+                }));
+            self.is_loading.set(true);
+            self.message_label.set_label("Loading ...");
+        }
+    }
+
+    impl WidgetImpl for MessageReply {}
+    impl BinImpl for MessageReply {}
+    impl MessageBaseImpl for MessageReply {}
+}
+
+glib::wrapper! {
+    pub(crate) struct MessageReply(ObjectSubclass<imp::MessageReply>)
+        @extends gtk::Widget, adw::Bin;
+}
+
+impl MessageReply {
+    pub(crate) async fn load_message(&self) {
+        let reply_id = self.imp().reply_id.get();
+        if reply_id == 0 {
+            self.update_message(None);
+        } else {
+            let chat = self.imp().chat.borrow().clone().unwrap();
+            let offline_message = chat.history().message_by_id(reply_id);
+            if offline_message.is_some() {
+                self.update_message(offline_message.as_ref());
+            } else {
+                spawn(clone!(@weak self as obj => async move {
+                    let chat = obj.imp().chat.borrow().clone().unwrap();
+                    let client_id = chat.session().client_id();
+                    let result = functions::get_message(chat.id(), reply_id, client_id).await.ok();
+                    let message = result.map(|result| {
+                     let EnumMessage::Message(message) = result;
+                     Message::new(message, &chat)
+                    });
+                    obj.update_message(message.as_ref());
+                }))
```

Comment:
Things have changes and this function should be implemented in the `Chat` instead. I think an appropriate name would be `request_message`. This should do the following:

1. Check if the message is present in the cache, and in that case just return that.
2. Otherwise call `functions::get_message` (asynchronously)
3. Store the returned message in the cache
4. Return the message